### DISCUSSION TITLE
Using Zluda for running Fooocus on AMD GPUs on Windows (instead of DirectML)

### DISCUSSION THREAD

**Discussion by Helikaon23s** [lllyasviel/Fooocus#2842] [Ideas]
Firstly, this guide is more for current users of ZLuda on SDNext or elsewhere (or new fork of Forge with ZLuda). The main errors in it not working in my observation is the setting of PATHs - it must be correct or it goes tits up pronto.
 I would advise using Notepad ++ for editing the python files, as it's far easier to ensure code is aligned (misalignment of code can also mess it up). 
Given the extra installs at the start it's not practical to fork this - I lack the python skills.

1.If ZLuda runs ok on SDNext following https://github.com/vladmandic/automatic/wiki/ZLUDA
  - you have installed Visual C++
  - you have installed the proper HIP for your card (important)
  - you have downloaded/unzipped ZLuda to a folder, mine ran with v3.5 but v3.71 is also stable https://github.com/lshqqytiger/ZLUDA/releases/
  - you have set the PATHs https://github.com/brknsoul/ROCmLibs/wiki/Adding-folders-to-PATH
  - note the section on disabling your iGpu
  
2.Download Fooocus & unzip https://github.com/lllyasviel/Fooocus

3.Navigate to "Fooocus/ldm_patched/modules/"

4.Open "model_management.py" in preferably notepad ++ 

5.Find this line (line 216)
`FORCE_FP32 = False`

6.Change line to `FORCE_FP32 = True` 
(this allows the ChatGPT2  / Fooocus Expansion V2 to work - you can if you want alternately add a startup argument for this)

 7.Find these lines (259-262)
```
try:
    print("Device:", get_torch_device_name(get_torch_device()))
except:
    print("Could not pick default device.")
```

8.Change the lines above by copy/pasting the following over them (this stop errors for cuda functions that don't work yet, enable bits that do and disables pytorch attention as ZLuda doesn't support it )
```
try:
    torch_device_name = get_torch_device_name(get_torch_device())

    if "[ZLUDA]" in torch_device_name:
        print("Detected ZLUDA, support for it is experimental and Fooocus may not work properly.")

        if torch.backends.cudnn.enabled:
            torch.backends.cudnn.enabled = False
            print("Disabling cuDNN because ZLUDA does currently not support it.")

        torch.backends.cuda.enable_flash_sdp(False)
        torch.backends.cuda.enable_math_sdp(True)
        torch.backends.cuda.enable_mem_efficient_sdp(False)

        if ENABLE_PYTORCH_ATTENTION:
            print("Disabling pytorch cross attention because ZLUDA does currently not support it.")
            ENABLE_PYTORCH_ATTENTION = False

    print("Device:", torch_device_name)
except:
    print("Could not pick default device.")
```

9.Install needed torch requirements after uninstalling the ones that don't work - open cmd window from main Fooocus folder with Imbeded folder 

```
.\python_embeded\python.exe -m pip uninstall torch torchvision -y
.\python_embeded\python.exe -m pip install torch==2.2.0 torchvision --index-url https://download.pytorch.org/whl/cu118
```

10.As per the older ZLuda instructions replace cublas64_11.dll, cusparse64_11.dll, and nvrtc64_112_0.dll in "python_embeded\Lib\site-packages\torch\lib" with the ones from the ZLUDA
 folder (replace and rename - checkout out the ZLuda wiki).
 
11. Start Fooocus by any of the 3 bat files, it should not need `--use-zluda` to start . It will download the files it wants for each of the 3 bat files as needed. 

12. PATHs checks, open a cmd window and type 
  zluda
  hipinfo

14. Start by running any of the startup up bat files (edited in as this was missing). No startup arguments needed - add Fooocus ones if you wish 

'ZLuda' should return text about its arguments
HIPinfo should return details of the installed HIP files - if it doesn't, you have an 'issue' with your PATH to them

-------------------------------------------------------------------------------------------------------------------------------------------------------

This was another proof on concept project for me - it's far slower (by its/s anyway or use Lightning models) than SDNext on my 7900xtx (SDNext 7.5its/s & Fooocus 1.2 - 2.2 its/s depending on speed setting). 

**Speed Increase**
Use Turbo, Lightning etc models 
On step 6, leave it at fp16 BUT disable Fooocus Expansion V2 (or it crashes)

ALL credit and kudos to  Vosen, lshqqytiger, BrknSoul & LeagueRaNi

**Comment by Dog-brother0113** [comment]
Your article is simply amazing. I am currently researching how Zluda works in various software I use, and you are like the lighthouse of night navigation. Thank you.

**Reply by Helikaon23s** [reply]
You are welcome.

**Comment by wtnwin** [comment]
@Grey3016 OSError: [WinError 126] The specified module could not be found. Error loading "E:\Fooocus\python_embeded\lib\site-packages\torch\lib\caffe2_nvrtc.dll" or one of its dependencies.
on 6700XT GPU while sd.next does work as instrunction followed.

plus zluda folder only has cublas.dll, cusparse.dll, and nvrtc64.dll, when these replace cublas64_11.dll, cusparse64_11.dll, and nvrtc64_112_0.dll in "python_embeded\Lib\site-packages\torch\lib" , do I need to rename change ublas.dll, cusparse.dll, and nvrtc64.dll to cublas64_11.dll, cusparse64_11.dll, and nvrtc64_112_0.dll in "python_embeded\Lib\site-packages\torch\lib"?

**Reply by Helikaon23s** [reply]
The Caffe error appears to be a reference to something not installed properly - Paths or HIP needs reinstalling. 
1. Add the python_embeded\Scripts to Path and see if that fixes it. 
2. Reinstall HIP libs and any specific to your card as required by SDnext wiki

**Reply by Helikaon23s** [reply]
It looks like you have the wrong ZLuda zip downloaded, after unzipping the folder should have about 8-10 files as I recall. Try downloading v3.5 which definitely works with Fooocus (I'm away from my pc for a while).
And yes to your final question - same files as per SDnext.

**Reply by wtnwin** [reply]
Thx for help, I did rename change cublas.dll, cusparse.dll, and nvrtc64.dll to cublas64_11.dll, cusparse64_11.dll, and nvrtc64_112_0.dll in "python_embeded\Lib\site-packages\torch\lib" to resolve problem. 

After success in running run.bat, should I use --use-zluda as arg to run run.bat to allow zluda to work? 

How to share the model with sd.next? so both sd UI can share the same model.

**Reply by Helikaon23s** [reply]
The --use-zluda argyment is not needed for this, the original method did but this version allows the usage of other Fooocus startup arguments - as a side note, I couldn't get it to work quicker with the ones I tried. I tried as many of the tools (advanced menus) as I could and it all worked (slow but it works). 

Sharing folders appears to be a bit of saga, you can symlink or someone says how to do it here https://github.com/lllyasviel/Fooocus/discussions/236

**Reply by wtnwin** [reply]
loading in lowvram mode 6935.72490978241
[Fooocus Model Management] Moving model(s) has taken 25.50 seconds
  0%|
...
    return torch.nn.functional.linear(input, weight, bias)
RuntimeError: CUDA error: CUBLAS_STATUS_NOT_SUPPORTED when calling `cublasSgemm( handle, opa, opb, m, n, k, &alpha, a, lda, b, ldb, &beta, c, ldc)`
Total time: 229.42 seconds

Once it takes full ram and varam, it switches to lowvram mode, then CUDA error above happens.

**Reply by Helikaon23s** [reply]
Ahh, I have a 7900, so it wouldn't have triggered the lowvram for me. I don't know if there are other arguments that could help with this, trying out ZLuda 3.71 might also help - I can't force lower vram to test it sorry.

**Comment by ghost** [comment]
I got this:
> RuntimeError: Found no NVIDIA driver on your system. Please check that you have an NVIDIA GPU and installed a driver from http://www.nvidia.com/Download/index.aspx

**Reply by Helikaon23s** [reply]
I can only guess that you did something out of sequence / missed something or your HIP needs a reinstall or your Paths are incorrect - this is why there's a reference to needing a working ZLuda copy of sdnext as a prerequisite.

**Reply by Helikaon23s** [reply]
I tested these instructions 5 times, they work, unless Fooocus has been updated.

**Reply by ghost** [reply]
@Grey3016 Okay, now it works, but freeze on step "Preparing Fooocus text #1"

**Reply by ghost** [reply]
> rocBLAS error: Cannot read C:\Program Files\AMD\ROCm\5.7\bin\/rocblas/library/TensileLibrary.dat: No such file or directory for GPU arch : gfx1036

> rocBLAS error: Could not initialize Tensile host:
regex_error(error_backref): The expression contained an invalid back reference.

**Reply by ghost** [reply]
Other Rocm projects are working (ex. koboldcpp_rocm), and zluda too (rendering in blender with cuda on AMD gpu). 
Only this doesn't work.

**Reply by Helikaon23s** [reply]
You've asked several things here and given no gpu and sparse info, I can only say that it definitely works for newer AMD cards, if you are using something old, then you must ensure you have the correct HIP files installed & your AMD drivers are up to date - 
1. Freezing fault > Did you do 'FORCE_FP32 = True' (step 6) ? 
2. The ROCBlas error > does that file exist (TensileLibrary.dat) ? if it doesn't, reinstall the HIP files, if it does, it implies your Paths aren't set correctly .
3. Try this, at a cmd, enter -
```
ZLuda 
Hipinfo
```
Both should return something. Returned info on each would imply that Paths have been set correctly.

**Reply by ghost** [reply]
@Grey3016 I have 7900 XTX and did install the latest ROCm HIP SDK (5.7.1).
I had enabled APU, that's why it showed gfx1036, but now I disabled it, but the problem persists.

**Reply by ghost** [reply]
Update: 
My `model_management.py` file was automatically reverted after step #9 for some reason.

**Reply by ghost** [reply]
@Grey3016 now it's working. Thank you for your assistance.

**Comment by ghost** [comment]
I tested twice: the code from steps 5 - 8 is reverted after performing step 9. It's need to change the order.

**Reply by Helikaon23s** [reply]
That's the idea, it automatically installed the latest torch for cuda (when you installed Fooocus) and this guide works by you manually uninstalling it and installing the ZLuda compatible torch (step 9). 
That's the reason every step is numbered, this isn't a guide for those that decide they're going to do their own thing. Don't run it between steps, step 9 only changes files inside the embeded folder. 
 As I said it works if the steps are followed.

**Reply by ghost** [reply]
As I said - I followed these steps in exact order and the code was reverted after the execution of step 9.

**Reply by Helikaon23s** [reply]
Whatever you have done is outside of the scope of the steps I have provided. I have asked you to carry out checks on which you provided no answers, I will have to terminate this discussion as I'm banging my head on the wall.

**Comment by lgwjames** [comment]
i spent close to 3 hours doing this on my windows bootcamp and first the run bat just opened and closed, then i go tsomething to do something but loads of errors and just closed

**Reply by ghost** [reply]
Here is the updated instruction (these steps worked for me):

### Requirements: 

1. you have installed Visual C++
2. you have installed the proper HIP for your card (https://www.amd.com/en/developer/resources/rocm-hub/hip-sdk.html)
3. you have downloaded and extract ZLuda to a folder you want (https://github.com/lshqqytiger/ZLUDA/releases/)
4. add this folder to the paths (environment variable)
5. disable your APU (iGpu) if you have it in your BIOS

### Steps

1. PATHs checks. open a cmd window and type
  - `zluda` (should print something like `Required positional arguments not provided`)
  - `hipinfo` (should print HIP SDK specs)
2. Download Fooocus & unzip (https://github.com/lllyasviel/Fooocus)
3. Install needed torch requirements after uninstalling the ones that don't work. Open `cmd.exe` window from main Fooocus folder with embedded folder

`.\python_embeded\python.exe -m pip uninstall torch torchvision -y`
`.\python_embeded\python.exe -m pip install torch==2.2.0 torchvision --index-url https://download.pytorch.org/whl/cu118`

4. As per the older ZLuda instructions replace `cublas64_11.dll`, `cusparse64_11.dll`, and `nvrtc64_112_0.dll` in `python_embeded\Lib\site-packages\torch\lib` with the according ones from the ZLUDA folder (create copies, rename and replace - checkout out the ZLuda wiki).
5. Run Fooocus with `run.bat` and wait for its window to open. Cmd should print massage containing `App started successful.`
6. Close Fooocus window and cmd.
7. Navigate to `Fooocus/ldm_patched/modules/`
8. Open `model_management.py` with any code editor
9. Find this line (line 216)
```py
FORCE_FP32 = False
````
change it to 
```py
FORCE_FP32 = True
```
(this allows the ChatGPT2 / Fooocus Expansion V2 to work - you can if you want alternately add a startup argument for this)

10. Find these lines (259-262)
```python
try:
    print("Device:", get_torch_device_name(get_torch_device()))
except:
    print("Could not pick default device.")
```
And replace them with the following code: 
```python
try:
    torch_device_name = get_torch_device_name(get_torch_device())

    if "[ZLUDA]" in torch_device_name:
        print("Detected ZLUDA, support for it is experimental and Fooocus may not work properly.")

        if torch.backends.cudnn.enabled:
            torch.backends.cudnn.enabled = False
            print("Disabling cuDNN because ZLUDA does currently not support it.")

        torch.backends.cuda.enable_flash_sdp(False)
        torch.backends.cuda.enable_math_sdp(True)
        torch.backends.cuda.enable_mem_efficient_sdp(False)

        if ENABLE_PYTORCH_ATTENTION:
            print("Disabling pytorch cross attention because ZLUDA does currently not support it.")
            ENABLE_PYTORCH_ATTENTION = False

    print("Device:", torch_device_name)
except:
    print("Could not pick default device.")
```
(this will stop errors from CUDA functions that don't work yet. Enables parts that do, and disables the pytorch attention as ZLuda doesn't support it )

11. Start Fooocus by any of the 3 bat files. (It should not need --use-zluda to start . It will download the files it wants for each of the 3 bat files as needed).

**Reply by Selos86** [reply]
Followed this instruction and had different results:

On the day performing the changes and starting for the first time I ended up with an error I unfortunaely did not make a screenshot of, but it was something like "GPU(0), falling back on CPU" but then it did nothing anymore.

Second day it was just doing nothing for over 1 1/2 hours after i started and entered a prompt (suprisingly the error from day one did not occur)

Today it started fine, loaded the model but then stopped at "preparing prompt" step and after approx. 5 minutes ended up with "Exception Code: 0xC0000094".
Also there are several dozens of lines similar to the following:

0x00007FFF760E45D6, _<PATH to Fooocus>_\python310.dll(0x00007FFF76060000) + 0x845D6 byte(s), PyType_IsSubtype() + 0x6FE byte(s)
0x00007FFF7607F33C, _<PATH to Fooocus>_\python310.dll(0x00007FFF76060000) + 0x1F33C byte(s), _PyEval_EvalFrameDefault() + 0x3F1C byte(s)
0x00007FFF7607A4B7, _<PATH to Fooocus>_\python310.dll(0x00007FFF76060000) + 0x1A4B7 byte(s), _PyFunction_Vectorcall() + 0x87 byte(s)
0x00007FFF760A7763, _<PATH to Fooocus>_\python310.dll(0x00007FFF76060000) + 0x47763 byte(s), _PyObject_FastCallDictTstate() + 0x6B byte(s)

Does anyone have an idea what I could to to get it working?

I am using a Radeon RX 6800 and a Ryzen 7 5800X

Thanks in advance!

**Reply by MarkovicM1** [reply]
Had the same situation, didn't change anything.
Basically redo steps 9 and 10. For what ever reason Fooocus/ldm_patched/modules/model_management.py reverted to the original.

Hope that helps

**Reply by Selos86** [reply]
Thanks for your feedback!

Just checked, the entries are still as suggested for using Zluda, but it still produces the above mentioned errors.

The error "Exception Code: 0xC0000094" seems to hint for some division by zero within the code of the program causing the error.

**Reply by chejo** [reply]
> Followed this instruction and had different results:
> 
> On the day performing the changes and starting for the first time I ended up with an error I unfortunaely did not make a screenshot of, but it was something like "GPU(0), falling back on CPU" but then it did nothing anymore.
> 
> Second day it was just doing nothing for over 1 1/2 hours after i started and entered a prompt (suprisingly the error from day one did not occur)
> 
> Today it started fine, loaded the model but then stopped at "preparing prompt" step and after approx. 5 minutes ended up with "Exception Code: 0xC0000094". Also there are several dozens of lines similar to the following:
> 
> 0x00007FFF760E45D6, \python310.dll(0x00007FFF76060000) + 0x845D6 byte(s), PyType_IsSubtype() + 0x6FE byte(s) 0x00007FFF7607F33C, \python310.dll(0x00007FFF76060000) + 0x1F33C byte(s), _PyEval_EvalFrameDefault() + 0x3F1C byte(s) 0x00007FFF7607A4B7, \python310.dll(0x00007FFF76060000) + 0x1A4B7 byte(s), _PyFunction_Vectorcall() + 0x87 byte(s) 0x00007FFF760A7763, \python310.dll(0x00007FFF76060000) + 0x47763 byte(s), _PyObject_FastCallDictTstate() + 0x6B byte(s)
> 
> Does anyone have an idea what I could to to get it working?
> 
> I am using a Radeon RX 6800 and a Ryzen 7 5800X
> 
> Thanks in advance!

That happened to me because fooocus has auto updated. You need to change the model_management.py file again

**Reply by Selos86** [reply]
It indeed reverted to the initial file content. I edited the file again and started it and the same errors still occur despite the model_management.py having the Zluda-conform content.

Does anyone have any further ideas?

**Reply by Selos86** [reply]
Ok, the above errors obviously occured due to Fooocus V2 being checked in the styles tab.

But now there is another error occuring when it tries to start image generation after all models are loaded:

```
[Fooocus] Preparing task 1/2 ...
[Sampler] refiner_swap_method = joint
[Sampler] sigma_min = 0.0291671771556139, sigma_max = 14.614643096923828
Traceback (most recent call last):
  File "D:\KI\Fooocus\Fooocus\modules\async_worker.py", line 998, in worker
    handler(task)
  File "D:\KI\Fooocus\python_embeded\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "D:\KI\Fooocus\python_embeded\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "D:\KI\Fooocus\Fooocus\modules\async_worker.py", line 888, in handler
    imgs = pipeline.process_diffusion(
  File "D:\KI\Fooocus\python_embeded\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "D:\KI\Fooocus\python_embeded\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "D:\KI\Fooocus\Fooocus\modules\default_pipeline.py", line 372, in process_diffusion
    modules.patch.BrownianTreeNoiseSamplerPatched.global_init(
  File "D:\KI\Fooocus\Fooocus\modules\patch.py", line 198, in global_init
    BrownianTreeNoiseSamplerPatched.tree = BatchedBrownianTree(x, t0, t1, seed, cpu=cpu)
  File "D:\KI\Fooocus\Fooocus\ldm_patched\k_diffusion\sampling.py", line 85, in __init__
    self.trees = [torchsde.BrownianTree(t0, w0, t1, entropy=s, **kwargs) for s in seed]
  File "D:\KI\Fooocus\Fooocus\ldm_patched\k_diffusion\sampling.py", line 85, in <listcomp>
    self.trees = [torchsde.BrownianTree(t0, w0, t1, entropy=s, **kwargs) for s in seed]
  File "D:\KI\Fooocus\python_embeded\lib\site-packages\torchsde\_brownian\derived.py", line 155, in __init__
    self._interval = brownian_interval.BrownianInterval(t0=t0,
  File "D:\KI\Fooocus\python_embeded\lib\site-packages\torchsde\_brownian\brownian_interval.py", line 540, in __init__
    W = self._randn(initial_W_seed) * math.sqrt(t1 - t0)
  File "D:\KI\Fooocus\python_embeded\lib\site-packages\torchsde\_brownian\brownian_interval.py", line 234, in _randn
    return _randn(size, self._top._dtype, self._top._device, seed)
  File "D:\KI\Fooocus\python_embeded\lib\site-packages\torchsde\_brownian\brownian_interval.py", line 33, in _randn
    return torch.randn(size, dtype=dtype, device=device, generator=generator)
**RuntimeError: Expected a 'cuda' device type for generator but found 'cpu'**
Total time: 368.88 seconds

```
Running "zluda --help" in the commandline returns the following, so I assume registering Zluda to PATH was successful:
```

C:\>zluda --help
Usage: zluda <exe> [<args...>] [--nccl <nccl>] [--nvrtc <nvrtc>] [--nvcuda <nvcuda>] [--nvml <nvml>] [--nvapi <nvapi>] [--nvoptix <nvoptix>]

Launch application with custom CUDA libraries

Positional Arguments:
  exe               executable to be injected with custom CUDA libraries
  args              arguments to the executable

Options:
  --nccl            DLL to be injected instead of system nccl.dll. If not
                    provided zluda, will use nccl.dll from its own directory
  --nvrtc           DLL to be injected instead of system nvrtc.dll. If not
                    provided zluda, will use nvrtc.dll from its own directory
  --nvcuda          DLL to be injected instead of system nvcuda.dll. If not
                    provided zluda, will use nvcuda.dll from its own directory
  --nvml            DLL to be injected instead of system nvml.dll. If not
                    provided zluda, will use nvml.dll from its own directory
  --nvapi           DLL to be injected instead of system nvapi64.dll. If not
                    provided, no injection will take place
  --nvoptix         DLL to be injected instead of system nvoptix.dll. If not
                    provided, no injection will take place
  --help            display usage information
```

Hipinfo also returns several information about my graphics card so I guess this is also fine.


After starting Fooocus with zluda, it states in the dialogue:

```
Detected ZLUDA, support for it is experimental and Fooocus may not work properly.
Disabling cuDNN because ZLUDA does currently not support it.
Disabling pytorch cross attention because ZLUDA does currently not support it.
Device: cuda:0 AMD Radeon RX 6800 [ZLUDA] : native
```

**Reply by Helikaon23s** [reply]
> Ok, the above errors obviously occured due to Fooocus V2 being checked in the styles tab.
> 
> But now there is another error occuring when it tries to start image generation after all models are loaded:
> 
> ```
> [Fooocus] Preparing task 1/2 ...
> [Sampler] refiner_swap_method = joint
> [Sampler] sigma_min = 0.0291671771556139, sigma_max = 14.614643096923828
> Traceback (most recent call last):
>   File "D:\KI\Fooocus\Fooocus\modules\async_worker.py", line 998, in worker
>     handler(task)
>   File "D:\KI\Fooocus\python_embeded\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
>     return func(*args, **kwargs)
>   File "D:\KI\Fooocus\python_embeded\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
>     return func(*args, **kwargs)
>   File "D:\KI\Fooocus\Fooocus\modules\async_worker.py", line 888, in handler
>     imgs = pipeline.process_diffusion(
>   File "D:\KI\Fooocus\python_embeded\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
>     return func(*args, **kwargs)
>   File "D:\KI\Fooocus\python_embeded\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
>     return func(*args, **kwargs)
>   File "D:\KI\Fooocus\Fooocus\modules\default_pipeline.py", line 372, in process_diffusion
>     modules.patch.BrownianTreeNoiseSamplerPatched.global_init(
>   File "D:\KI\Fooocus\Fooocus\modules\patch.py", line 198, in global_init
>     BrownianTreeNoiseSamplerPatched.tree = BatchedBrownianTree(x, t0, t1, seed, cpu=cpu)
>   File "D:\KI\Fooocus\Fooocus\ldm_patched\k_diffusion\sampling.py", line 85, in __init__
>     self.trees = [torchsde.BrownianTree(t0, w0, t1, entropy=s, **kwargs) for s in seed]
>   File "D:\KI\Fooocus\Fooocus\ldm_patched\k_diffusion\sampling.py", line 85, in <listcomp>
>     self.trees = [torchsde.BrownianTree(t0, w0, t1, entropy=s, **kwargs) for s in seed]
>   File "D:\KI\Fooocus\python_embeded\lib\site-packages\torchsde\_brownian\derived.py", line 155, in __init__
>     self._interval = brownian_interval.BrownianInterval(t0=t0,
>   File "D:\KI\Fooocus\python_embeded\lib\site-packages\torchsde\_brownian\brownian_interval.py", line 540, in __init__
>     W = self._randn(initial_W_seed) * math.sqrt(t1 - t0)
>   File "D:\KI\Fooocus\python_embeded\lib\site-packages\torchsde\_brownian\brownian_interval.py", line 234, in _randn
>     return _randn(size, self._top._dtype, self._top._device, seed)
>   File "D:\KI\Fooocus\python_embeded\lib\site-packages\torchsde\_brownian\brownian_interval.py", line 33, in _randn
>     return torch.randn(size, dtype=dtype, device=device, generator=generator)
> **RuntimeError: Expected a 'cuda' device type for generator but found 'cpu'**
> Total time: 368.88 seconds

> ```

Can you confirm you have ZLuda working with SDNext ? 
Can you confirm you have done EVERY step, in the order they are in ? 

I would guess that either the torch wasn't installed or the the 3 files that need to be changed/renamed were not. SDNext now does most of the original steps automatically with scripts, so the instructions will change. 
If you are getting errors due to Expansion V2 being turned on means you haven't carried out steps 5/6. I don't know what to say other than it works if the steps are followed.
Mmmm... check that the copy/pasted code in model_management aligns exactly as my original post - no extra spaces, code aligns with code above and below. This is why I mentioned Notepad++ instead of any old text editor, N++ shows the alignment of code blocks .

**Reply by Selos86** [reply]
I followed the instructions step by step. I renamed the 3 files as suggested (see screenshot) and also steh code in model_management.py is as suggested (see other screenshot)
SD.Next I wanted to try but it does not seem to come with an embedded python and I don't hava a general python installation on this machine.

![ZLUDA](https://github.com/lllyasviel/Fooocus/assets/148383763/9b9aaee3-61fc-44ba-8a72-51dd67c979b8)

![ZLUDA2 jpg](https://github.com/lllyasviel/Fooocus/assets/148383763/743744a0-9614-42cf-bdd6-f7a8e853a44a)

**Comment by cavalia88** [comment]
I tried following the instructions but could not get Fooocus to work with zluda. I have zluda working with Automatic1111 and SD_NEXT, so most of the required stuff should already be there. 

OSError: [WinError 126] The specified module could not be found. Error loading "E:\Fooocus\python_embeded\lib\site-packages\torch\lib\caffe2_nvrtc.dll" or one of its dependencies.

**Reply by Helikaon23s** [reply]
> I tried following the instructions but could not get Fooocus to work with zluda. I have zluda working with Automatic1111 and SD_NEXT, so most of the required stuff should already be there. 
> 
> 
> 
> OSError: [WinError 126] The specified module could not be found. Error loading "E:\Fooocus\python_embeded\lib\site-packages\torch\lib\caffe2_nvrtc.dll" or one of its dependencies.

Try reinstalling HIP first, for some reason it can go tits up.
Then edit your Paths and add "E:\Fooocus\python_embeded\lib\site-packages\torch\lib" to your paths.

**Reply by cavalia88** [reply]
Still does not work. Added the directory to my paths, but still encounter the same error

**Reply by Helikaon23s** [reply]
Did you reinstall HIP and check it installs the Paths correctly? Also, type into a cmd - 
ZLuda
Hipinfo

Do they both return something?

**Reply by cavalia88** [reply]
I found the issue. As you suspected, it was the paths. Zluda path was fine, the problem was the Hipinfo one. It was not very clear what was required to be added for the Hipinfo path. Came across another page which provided the answer: https://github.com/CS1o/Stable-Diffusion-Info/wiki/Installation-Guides

Add the C:\ZLUDA folder and %HIP_PATH%bin to your PATH in the System Variables.

Managed to load fooocus. Trying to generate the first image but encountered this error after generating for some time: 
RuntimeError: cuDNN error: CUDNN_STATUS_INTERNAL_ERROR
Total time: 379.57 seconds

**Reply by Helikaon23s** [reply]
> I found the issue. As you suspected, it was the paths. Zluda path was fine, the problem was the Hipinfo one. It was not very clear what was required to be added for the Hipinfo path. Came across another page which provided the answer: https://github.com/CS1o/Stable-Diffusion-Info/wiki/Installation-Guides
> 
> Add the C:\ZLUDA folder and %HIP_PATH%bin to your PATH in the System Variables.
> 
> Managed to load fooocus. Trying to generate the first image but encountered this error after generating for some time: RuntimeError: cuDNN error: CUDNN_STATUS_INTERNAL_ERROR Total time: 379.57 seconds

The idea of this guide was for ppl who already had it installed (ie Paths set and HIP installed ok first. If it all boots fine - I would suspect you didn't do Step 5/6 and the line is set to False (or it changed back on an update). Change it to True and retry.

**Reply by cavalia88** [reply]
I double checked and confirm that "FORCE_FP32 = True". Still getting the same error. Not sure what else i can try to get it working...probably have to fall back on automatic1111 and sd_NEXT

**Reply by Helikaon23s** [reply]
Turn off the Expansion V2 filter before you start the first generation, in case that is playing up.

**Reply by cavalia88** [reply]
Yes, I have already unchecked "Fooocus V2" in the style tab before running the first generation. Still getting the same error

**Reply by Helikaon23s** [reply]
Recheck the copy & pasted code in model_managment - check the code is aligned ie no extra spaces or code not in line with code above or below (this is why I advocated for Notepad+ not any old text editor, it shows the alignment of code).  
Other than that, go back and recheck each step. I know it will work on a 6800.

**Reply by cavalia88** [reply]
Finally managed to get it to work on my 7900XTX. The issue was that the code I pasted previously had reverted back to its original form. I assume this was due to the upgrading of one of the components. So each time there is an upgrade, we will need to copy/paste the code over one more time. Nonetheless, thank you for your patience and advise along the way.

**Comment by urostrickovic** [comment]
![image_2024-07-27_151947660](https://github.com/user-attachments/assets/658b8a83-1920-4a86-a5a5-5126dfd85753)
Did anyone solve this problem: 
RuntimeError: CUDA error: CUBLAS_STATUS_NOT_SUPPORTED when calling `cublasSgemm( handle, opa, opb, m, n, k, &alpha, a, lda, b, ldb, &beta, c, ldc)`

**Reply by Helikaon23s** [reply]
> ![image_2024-07-27_151947660](https://github.com/user-attachments/assets/658b8a83-1920-4a86-a5a5-5126dfd85753)
> 
> Did anyone solve this problem: 
> 
> RuntimeError: CUDA error: CUBLAS_STATUS_NOT_SUPPORTED when calling `cublasSgemm( handle, opa, opb, m, n, k, &alpha, a, lda, b, ldb, &beta, c, ldc)`
> 
> 
> 
> 

I think (seems a long time ago now), this issue is caused by the wrong version of ZLuda being used. At the time of writing this, I recall using v1.5 (my HD is now wiped) as it was a datum point (this ZLuda version worked with Fooocus, SDNext and for that version of Forge - all of these have been updated since)

**Comment by Vitaliy290590** [comment]
I hope this publication is still alive.
I did everything step by step. zluda, hipinfo commands return arguments. 
When Fooocus starts, the message "pause" appears. And that's it. virtual memory is 60 gigabytes

**Reply by Helikaon23s** [reply]
This guide has been superseded by a dedicated Zluda Forge release - do a search on Reddit SD reddit for AMD / Flux and a guide I wrote (in my posts) will give you the url (and other dedicated Zluda urls for Comfy)

**Reply by Vitaliy290590** [reply]
thank you for your work

**Comment by hltzngnn** [comment]
I am running the Fooocus application through Google Colab, but the input image feature has not been working for a while.

**Reply by Helikaon23s** [reply]
This guide has been superseded by a dedicated Zluda Forge release - do a search on Reddit SD reddit for AMD / Flux and a guide I wrote (in my posts) will give you the url (and other dedicated Zluda urls for Comfy)

**Comment by XDM-Inc** [comment]
might you know how to get this working on Linux by chance? i downloaded stable diffusion NEXT with Zluda working  and its pretty fast but i have no idea how to port it to fooocus

**Reply by Helikaon23s** [reply]
No

**Comment by XDM-Inc** [comment]
@Grey3016 you say that This guide has been superseded by a dedicated Zluda Forge release but that guide you are referring to is for stock stable diffusion and not fooocus as far as i can see if [this](https://www.reddit.com/r/StableDiffusion/comments/1brm3rc/sd_forge_for_amd_gpus_with_zluda_on_windows_v10/) is what your talking about. if thats not it then what page might you be talking about?

**Reply by Helikaon23s** [reply]
This is 7 months old, I wrote it as a proof of concept that it worked with my other amd gpu. Fooocus has been updated since then and I've no idea about any of it (as my primary gpu is an nvidia one). I got myself mixed up with comments in Reddit (between Fooocus and Forge) and here and there isn't a dedicated zluda branch afaik, the Zluda branch of Comfy (as I recall) does have great comments within its Github page that show how that person converted it to Zluda. I won't be revisiting this project to update or check it, sorry.

**Comment by CS1o** [comment]
Fooocus with ZLUDA still works. I updated my Guide for the latest Version.
You can find it here with all the other Webui Guides:
https://github.com/CS1o/Stable-Diffusion-Info/wiki/Webui-Installation-Guides
